main/src/main/java/cgeo/geocaching/models/Geocache.java

```diff
@@ -1591,7 +1594,13 @@ public boolean addOrChangeWaypoint(final Waypoint waypoint, final boolean saveTo
             // when waypoint was edited, finalDefined may have changed
             resetFinalDefined();
         }
-        return saveToDatabase && DataStore.saveWaypoint(waypoint.getId(), geocode, waypoint);
+
+        boolean savedToDatabase = false;
+        if (saveToDatabase) {
+            DataStore.saveFinalDefinedStatus(this);
+            savedToDatabase = DataStore.saveWaypoint(waypoint.getId(), geocode, waypoint);
```

Comment:
Maybe `savefinaldefinedstatus` should be called from inside `savewaypoint`?

Reply:
Would be possible, but then I need to pass the cache / finalDefined...
If that would be fine...

Reply:
Isn't the check for finaldefined necessary for every "save waypoint" call anyway? Because every change of any waypoint ma lead to a change of this flag

Reply:
Eg if someone changes the type of a waypoint from "final" to "puzzle" or vice versa

Reply:
I think, this is covered by addOrChangrWaypoint
But I would like to have it together in one method to avoid discrepancy

Reply:
Merged the new function into saveWaypoint